src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java

```diff
@@ -0,0 +1,330 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQuerySQLEngine.class);
+
+  public static final String NAME = "BigQueryPushdownEngine";
+
+  private final BigQuerySQLEngineConfig config;
+  private BigQuery bigQuery;
+  private Storage storage;
+  private Configuration configuration;
+  private String project;
+  private String dataset;
+  private String bucket;
+  private String runId;
+  private Map<String, String> tableNames;
+  private Map<String, BigQuerySQLDataset> datasets;
+
+  @SuppressWarnings("unused")
+  public BigQuerySQLEngine(BigQuerySQLEngineConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+  }
+
+  @Override
+  public void prepareRun(RuntimeContext context) throws Exception {
+    super.prepareRun(context);
+
+    runId = BigQuerySQLEngineUtils.getIdentifier();
+    tableNames = new HashMap<>();
+    datasets = new HashMap<>();
+
+    String serviceAccount = config.getServiceAccount();
+    Credentials credentials = serviceAccount == null ?
+      null : GCPUtils.loadServiceAccountCredentials(serviceAccount, config.isServiceAccountFilePath());
+    project = config.getProject();
+    dataset = config.getDataset();
+    bucket = config.getBucket() != null ? config.getBucket() : "bqpushdown-" + runId;
+
+    // Initialize BQ and GCS clients.
+    bigQuery = GCPUtils.getBigQuery(project, credentials);
+    storage = GCPUtils.getStorage(project, credentials);
+
+    String cmekKey = context.getRuntimeArguments().get(GCPUtils.CMEK_KEY);
+    configuration = BigQueryUtil.getBigQueryConfig(config.getServiceAccount(), config.getProject(),
+                                                   cmekKey, config.getServiceAccountType());
+    BigQuerySinkUtils.createResources(bigQuery, storage, dataset, bucket, config.getLocation(), cmekKey);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, RuntimeContext context) {
+    super.onRunFinish(succeeded, context);
+  }
+
+  @Override
+  public SQLPushDataset<StructuredRecord, StructuredRecord, NullWritable> getPushProvider(SQLPushRequest sqlPushRequest)
+    throws SQLEngineException {
+    try {
+      BigQueryPushDataset pushDataset =
+        BigQueryPushDataset.getInstance(sqlPushRequest,
+                                        configuration,
+                                        bigQuery,
+                                        project,
+                                        dataset,
+                                        bucket,
+                                        runId);
+
+      LOG.info("Executing Push operation for dataset {} stored in table {}",
+               sqlPushRequest.getDatasetName(),
+               pushDataset.getBigQueryTableName());
+
+      datasets.put(sqlPushRequest.getDatasetName(), pushDataset);
```

Comment:
it would be better if the engine didn't have to keep track of state like this (it means every implementation would have to do something similar).

It seems like you don't actually need to keep track though, unless I'm missing something? For pulling and cleanup, the source of truth is always if the table exists in BQ.

Reply:
hm I see, let's keep this for now.

Is there a TTL being set on tables that are created? We also need to consider situations where the job is killed/dies without cleanup being run. Can do this in a follow-up PR if not there.

Reply:
We can add table TTL on table creation (using the table output format provider). I will add this in a later PR.